diff --git a/build.zig b/build.zig
index f9d861b19..88acdfca0 100644
--- a/build.zig
+++ b/build.zig
@@ -130,24 +130,18 @@ pub fn build(b: *std.Build) !void {
         }
     } else {
         // Libghostty
-        //
-        // Note: libghostty is not stable for general purpose use. It is used
-        // heavily by Ghostty on macOS but it isn't built to be reusable yet.
-        // As such, these build steps are lacking. For example, the Darwin
-        // build only produces an xcframework.
-
-        // We shouldn't have this guard but we don't currently
-        // build on macOS this way ironically so we need to fix that.
+        libghostty_static.installHeader();
+        libghostty_static.install("libghostty.a");
         if (!config.target.result.os.tag.isDarwin()) {
-            libghostty_shared.installHeader(); // Only need one header
             libghostty_shared.install("libghostty.so");
-            libghostty_static.install("libghostty.a");
         }
     }
 
     // macOS only artifacts. These will error if they're initialized for
     // other targets.
-    if (config.target.result.os.tag.isDarwin()) {
+    if (config.target.result.os.tag.isDarwin() and
+        (config.emit_xcframework or config.emit_macos_app))
+    {
         // Ghostty xcframework
         const xcframework = try buildpkg.GhosttyXCFramework.init(
             b,
@@ -164,17 +158,17 @@ pub fn build(b: *std.Build) !void {
         }
 
         // Ghostty macOS app
-        const macos_app = try buildpkg.GhosttyXcodebuild.init(
-            b,
-            &config,
-            .{
-                .xcframework = &xcframework,
-                .docs = &docs,
-                .i18n = if (i18n) |v| &v else null,
-                .resources = &resources,
-            },
-        );
         if (config.emit_macos_app) {
+            const macos_app = try buildpkg.GhosttyXcodebuild.init(
+                b,
+                &config,
+                .{
+                    .xcframework = &xcframework,
+                    .docs = &docs,
+                    .i18n = if (i18n) |v| &v else null,
+                    .resources = &resources,
+                },
+            );
             macos_app.install();
         }
     }
@@ -202,7 +196,7 @@ pub fn build(b: *std.Build) !void {
 
         // On macOS we can run the macOS app. For "run" we always force
         // a native-only build so that we can run as quickly as possible.
-        if (config.target.result.os.tag.isDarwin()) {
+        if (config.target.result.os.tag.isDarwin() and config.emit_macos_app) {
             const xcframework_native = try buildpkg.GhosttyXCFramework.init(
                 b,
                 &deps,
